[WIP] Add double free detection in thread cache for debug build

izaitsevfb opened the pull request:
Add new runtime option `debug_dblfree_cbin_scan_limit` that specifies the limit of for number of stack entries to scan in cache bit when trying to detect double free (debug build only).


### Context:

Double free is a common pattern to cause heap corruption. Typically we expect ASAN to catch these, however it's not always feasible because of the high CPU & memory overhead of ASAN. Thus we still rely on the debug build of jemalloc to catch things sometimes.

The thread cache (tcache) is critical to performance since it serves as the common and fast path for the majority of allocations. By default the debug build has tcache on, so that it's not too slow -- however the downside is, it could easily hide double frees. Currently the sanity checks may still discover double frees, but only down the arena path where the original troubled free is in the past.

The first step of the double free checking, is to scan the existing items within the tcache and see if there's any match.

interwq commented on include/jemalloc/internal/cache_bin.h:
```diff
@@ -436,6 +461,12 @@ cache_bin_dalloc_easy(cache_bin_t *bin, void *ptr) {
 	}
 
+        if (config_debug && opt_debug_dblfree_cbin_scan_limit != 0) {
+                if (cache_bin_validate_double_free(bin, ptr)) {
```
With the change mentioned above, this can just become `if (cache_bin_dalloc_safety_checks(...)) {`

so that the dalloc_easy impl isn't directly exposed to the details of the safety_checks.

izaitsevfb replied:
addressed in [a903674](https://github.com/jemalloc/jemalloc/pull/2310/commits/a903674db54ec8e8fc00b588e4e4279b0e0cd965)

interwq commented on include/jemalloc/internal/cache_bin.h:
```diff
@@ -427,6 +428,30 @@ cache_bin_full(cache_bin_t *bin) {
+ * Fires safety_check_fail if the ptr is found and returns false.
+ */
+JEMALLOC_ALWAYS_INLINE bool
+cache_bin_validate_double_free(cache_bin_t *bin, void *ptr) {
```
nit: let's name it `cache_bin_dalloc_safety_checks()`, and add the following logic at the beginning:

```
if (!config_debug || opt_debug_double_free_max_scan == 0) {
    return false;
}
```

izaitsevfb replied:
addressed in [a903674](https://github.com/jemalloc/jemalloc/pull/2310/commits/a903674db54ec8e8fc00b588e4e4279b0e0cd965)

interwq commented on include/jemalloc/internal/cache_bin.h:
```diff
@@ -427,6 +428,30 @@ cache_bin_full(cache_bin_t *bin) {
+                safety_check_fail(
+                    "Invalid deallocation detected: double free\n");
+                return true;
+        }
```
With the `max_scan == 0` check at the beginning, we can simplify above a bit:
```
 while (limit != 0 && (uint16_t)(uintptr_t)cur != bin->low_bits_full) {
    if (*cur == ptr) {
        ...
    }
    limit--;
    cur++;
}
```

izaitsevfb replied:
very elegant!

izaitsevfb replied:
addressed in [a903674](https://github.com/jemalloc/jemalloc/pull/2310/commits/a903674db54ec8e8fc00b588e4e4279b0e0cd965)

interwq commented on include/jemalloc/internal/cache_bin.h:
```diff
@@ -427,6 +428,30 @@ cache_bin_full(cache_bin_t *bin) {
+        // ptr is already present in the stack
+        if (limit != 0 && ((uint16_t)(uintptr_t)cur != bin->low_bits_full)) {
+                safety_check_fail(
+                    "Invalid deallocation detected: double free\n");
```
Let's also output the ptr value in the failure message.

izaitsevfb replied:
addressed in [a903674](https://github.com/jemalloc/jemalloc/pull/2310/commits/a903674db54ec8e8fc00b588e4e4279b0e0cd965)

interwq commented on include/jemalloc/internal/jemalloc_internal_externs.h:
```diff
@@ -34,6 +34,7 @@ extern malloc_init_t malloc_init_state;
 extern const char *zero_realloc_mode_names[];
 extern atomic_zu_t zero_realloc_count;
 extern bool opt_cache_oblivious;
+extern unsigned opt_debug_dblfree_cbin_scan_limit;
```
Maybe `opt_debug_double_free_max_scan`

izaitsevfb replied:
addressed in [5a2502a](https://github.com/jemalloc/jemalloc/pull/2310/commits/5a2502a2dd9fb3e401373e087983f334ed35dbc4)

interwq commented on src/jemalloc.c:
```diff
@@ -1420,6 +1422,9 @@ malloc_conf_init_helper(sc_data_t *sc_data, unsigned bin_shard_sizes[SC_NBINS],
 			    CONF_CHECK_MIN, CONF_CHECK_MAX, /* clip */ true)
+			CONF_HANDLE_UNSIGNED(opt_debug_dblfree_cbin_scan_limit,
+			    "debug_dblfree_cbin_scan_limit", 0, 1024,
+                            CONF_DONT_CHECK_MIN, CONF_CHECK_MAX, /* clip */ true)
```
No need to really limit the max, i.e. set it to UINT_MAX, and CONF_DONT_CHECK_MAX

If the user wants to set a max value to scan as many as possible, let's allow it.

izaitsevfb replied:
addressed in [5a2502a](https://github.com/jemalloc/jemalloc/pull/2310/commits/5a2502a2dd9fb3e401373e087983f334ed35dbc4)

interwq commented on src/jemalloc.c:
```diff
@@ -154,6 +154,8 @@ fxp_t		opt_narenas_ratio = FXP_INIT_INT(4);
 
 unsigned	ncpus;
 
+unsigned	opt_debug_dblfree_cbin_scan_limit;
```
For setting the default value, here you can do `unsigned opt_debug_double_free_max_scan = SAFETY_CHECK_DOUBLE_FREE_MAX_SCAN_DEFAULT;`
and then in `safety_check.h` have
```
#define SAFETY_CHECK_DOUBLE_FREE_MAX_SCAN_DEFAULT 32
```

Also let's add this in https://github.com/jemalloc/jemalloc/blob/dev/src/jemalloc.c#L1734
```
if (!config_debug) {
    opt_debug_double_free_max_scan = 0;
}
```

It doesn't really affect anything, but should make the stats output clearer (that it's disabled).

izaitsevfb replied:
addressed in [5a2502a](https://github.com/jemalloc/jemalloc/pull/2310/commits/5a2502a2dd9fb3e401373e087983f334ed35dbc4)

interwq commented on test/unit/double_free.c:
```diff
@@ -70,8 +70,30 @@ TEST_BEGIN(test_large_double_free_no_tcache) {
+TEST_BEGIN(test_double_free_tcache) {
+    test_skip_if(!config_debug);
+
+    test_large_double_free_pre();
```
nit: Let's remove the `large` part from the _pre and _post functions names. You will also want to skip the test if tcache is disabled: https://github.com/jemalloc/jemalloc/blob/d660683d3ddc2aaebf41a5662a6bc629be016e6d/test/unit/uaf.c#L51

izaitsevfb replied:
addressed in [4bc176c](https://github.com/jemalloc/jemalloc/pull/2310/commits/4bc176c3002fa751f267c7a420d38a8a2df2e712)

interwq commented on test/unit/double_free.c:
```diff
@@ -70,8 +70,30 @@ TEST_BEGIN(test_large_double_free_no_tcache) {
+    test_skip_if(!config_debug);
+
+    test_large_double_free_pre();
+    char *ptr = malloc(SC_NTINY);
```
malloc(1) is fine. SC_NTINY is actually the number of tiny size classes.

izaitsevfb replied:
addressed in [4bc176c](https://github.com/jemalloc/jemalloc/pull/2310/commits/4bc176c3002fa751f267c7a420d38a8a2df2e712)

interwq commented on test/unit/double_free.c:
```diff
@@ -70,8 +70,30 @@ TEST_BEGIN(test_large_double_free_no_tcache) {
 }
 TEST_END
 
+TEST_BEGIN(test_double_free_tcache) {
```
nit: The name `test_double_free_small_tcache`

izaitsevfb replied:
addressed in [4bc176c](https://github.com/jemalloc/jemalloc/pull/2310/commits/4bc176c3002fa751f267c7a420d38a8a2df2e712)

davidtgoldblatt commented:
I probably won't do an in-depth review, but one comment from a skim: I think this is a lot more useful/powerful if it's not debug-mode-only (e.g. what's the point if you have to re-build / re-link to get the checking; if you're rebuilding anyways, you might as well rebuild with ASAN and get way better checking.). We can get that for free by folding the option check into the fast-path check. So, the cache-bin code won't read the global directly; instead it'll get passed in as a static parameter down fast paths (and a dynamic one down slow paths).

The `opt.junk` and `opt.zero` handling is perhaps a good example of how to do that.

interwq commented:
@davidtgoldblatt agreed; that will be one of the followup items around this topic (the other being checking the slab bitmap in addition to scanning the stack).

interwq commented on include/jemalloc/internal/cache_bin.h:
```diff
@@ -436,6 +465,10 @@ cache_bin_dalloc_easy(cache_bin_t *bin, void *ptr) {
 		return false;
 	}
 
+        if (cache_bin_dalloc_safety_checks(bin, ptr)) {
```
nit: let's mark this `if (unlikely(...))`

izaitsevfb replied:
addressed in [60a3574](https://github.com/jemalloc/jemalloc/pull/2310/commits/60a357485181d0cb1f45c6c04a7641992e9178ff)

interwq commented on include/jemalloc/internal/cache_bin.h:
```diff
@@ -427,6 +428,34 @@ cache_bin_full(cache_bin_t *bin) {
+        void **cur = bin->stack_head;
+        unsigned limit = opt_debug_double_free_max_scan;
+
+        while (limit != 0 && (uint16_t)(uintptr_t)cur != bin->low_bits_empty) {
```
nit: Let's try optimizing a bit further -- for debug build it actually doesn't matter much; however we may add it to opt build later, also it can be a good example.
```
cache_bin_sz_t ncached = cache_bin_ncached_get_internal(bin, false);
unsigned max_scan = opt_debug_double_free_max_scan < ncached ? opt_debug_double_free_max_scan : ncached;

void **cur = bin->stack_head;
for (void **limit = cur + max_scan; cur < limit; cur++) {
    if (*cur == ptr) {
        ...
    }
}
```
This way we can avoid 1 branch check per iteration.

For the comments we always do `/* */`. It can be either one line, or multiple line format like you have for this function.

izaitsevfb replied:
addressed in [60a3574](https://github.com/jemalloc/jemalloc/pull/2310/commits/60a357485181d0cb1f45c6c04a7641992e9178ff)

interwq commented on test/unit/double_free.c:
```diff
@@ -44,15 +44,15 @@ TEST_BEGIN(test_large_double_free_tcache) {
 		fake_abort_called = true;
 	}
 	mallctl("thread.tcache.flush", NULL, NULL, NULL, 0);
```
Let's move this to `_post()`, given that it's needed in the new test as well.

izaitsevfb replied:
Do I understand correctly that it's ok to call `mallctl("thread.tcache.flush", NULL, NULL, NULL, 0);`  in `test_large_double_free_no_tcache` as well?
https://github.com/jemalloc/jemalloc/blob/4bc176c3002fa751f267c7a420d38a8a2df2e712/test/unit/double_free.c#L69

interwq replied:
Yes that's fine.

izaitsevfb replied:
addressed in [60a3574](https://github.com/jemalloc/jemalloc/pull/2310/commits/60a357485181d0cb1f45c6c04a7641992e9178ff)

interwq commented on test/unit/double_free.c:
```diff
@@ -66,12 +66,43 @@ TEST_BEGIN(test_large_double_free_no_tcache) {
 main(void) {
-	return test(test_large_double_free_no_tcache,
-	    test_large_double_free_tcache);
+        return test(test_large_double_free_no_tcache,
```
The indent change probably unintended? It should remain tabs.

For the new test below, the convention here is to add it in a separate line.

izaitsevfb replied:
I screwed up the indent style in my changes (used only spaces instead of tabs + spaces), I'll reformat and force-push everything to fix that.

izaitsevfb replied:
I went through all my changes and made sure that the indentation and comment style is correct:
https://github.com/jemalloc/jemalloc/compare/60a357485181d0cb1f45c6c04a7641992e9178ff..eb070173cceeaf545c6ffb1460facb3659b3ffb3

interwq commented:
Travis CI has a few failures but all on their side, none related to this diff.

izaitsevfb committed:
Add double free detection in thread cache for debug build

Add new runtime option `debug_double_free_max_scan` that specifies the max number of stack entries to scan in the cache bit when trying to detect the double free bug (currently debug build only).

interwq commented:
Tweaked a bit and merged as 36366f3c4c741723369853c923e56999716398fc with your authorship @izaitsevfb. Thank you for working on this!
